core/src/main/scala/kafka/admin/AdminClient.scala

```diff
@@ -52,15 +52,28 @@ class AdminClient(val time: Time,
                   val bootstrapBrokers: List[Node]) extends Logging {
 
   @volatile var running: Boolean = true
+  @volatile var nonRetriableException: Option[ApiException] = null
   val pendingFutures = new ConcurrentLinkedQueue[RequestFuture[ClientResponse]]()
 
   val networkThread = new KafkaThread("admin-client-network-thread", new Runnable {
     override def run() {
       try {
         while (running) {
+          var authenticationPerformed = true
           client.poll(Long.MaxValue)
+          bootstrapBrokers.foreach { broker =>
+            val apiException = client.authenticationFailed(broker)
+            if (apiException != null)
+              throw apiException
+            if (!client.isReady(broker))
+              authenticationPerformed = false
+          }
+          if (authenticationPerformed)
+            nonRetriableException = None
```

Comment:
I dont think we need to check or remember if authentication failed. Instead we should propagate AuthenticationExceptions from `AdminClient.sendAnyNode` to avoid retrying. I think this block of change can be removed.

Reply:
Thanks. I'll use your suggested approach.